### DISCUSSION TITLE
3D scene embed with PyVista (vtk.js)

### DISCUSSION THREAD

**Discussion by FelipeCybis** [quarto-dev/quarto-cli#7968] [Q&A] [answered]
### Description

Recently, PyVista introduced a (still in development but not bad) html exporter of its 3D scene. Basically it wraps everything in vtk.js and loads it in an offline viewer.

With `jupyter nbconvert` it works pretty well, but I did not have the same luck with quarto.

Anyways, here is the code I am using to test this ~(took out one back quote from the code cells for better rendering here)~:
````
---
title: HTML export
execute:
  enabled: true
embed-resources: false
jupyter: python3
---

```{python}
import pyvista
from ipywidgets import widgets
```

```{python}
widgets.IntSlider()
```

```{python}
pyvista.set_plot_theme('dark')
plotter = pyvista.Plotter()


grid = pyvista.ImageData(dimensions=(9, 9, 9))
grid['scalars'] = -grid.x
pl = pyvista.Plotter()
_ = pl.add_volume(grid, opacity='linear', mapper="smart")

pl.show(jupyter_backend="html")
```

## New section

```{python}
print("Hello")
```
````

The problem here is how the script inside the `ipywidgets` that is generated by `pl.show(jupyter_backend="html")` is decoded in `quarto`, because I can tweak the html after and make it work. For example there is one part of the script that should be `;\"><\/iframe>"`, but quarto generates `;\"&gt;"`. Then there are also some `</script>` that apparently should be `<\/script>`

I am not at all expert in javascript and html so I just back engineered the html outputed by `nbconvert` to make it work in the html generated by quarto.

What I actually want to know is where the change should be done and what change... Is it a quarto thing when it decodes the `ipywidget`? is it the string in the `ipywidget` that should be encoded differently? why `nbconvert` reads and decodes the string correctly and not quarto? I would love someone help on this :)

Here are all the files I am using
https://github.com/FelipeCybis/trame_quarto_example

Also I am using quarto `1.4.514`.

Thanks in advance and thanks for the amazing quarto :)

**Comment by mcanouil** [comment]
Don't remove backticks, add more on the outer code cell.

**Reply by FelipeCybis** [reply]
There you go, thanks

**Comment by mcanouil** [comment]
Using latest pre-release version, latest Python and latest modules I have no issues.
Note that it seems, the module requires a running Python server, so you might not be able to publish/share HTML files.


https://github.com/quarto-dev/quarto-cli/assets/8896044/d9396267-86f8-46ab-9177-512fe6157ba3

**Reply by FelipeCybis** [reply]
Hi @mcanouil ! Sorry for the late answer.

Thanks for your answer. Yes, I can reproduce your example, but you are using `jupyter_backend="trame"` (the default when `pl.show()`) so yes, a trame server is created and all that.

But the html export in trame and pyvista was developped exactly to work in standalone html exports. That is what I am trying to achieve, which is the behaviour I get with `jupyter nbconvert --to html quarto_example.ipynb --execute`, for example. If you go to https://github.com/FelipeCybis/trame_quarto_example and download the `pyvista_nbconvert.html` you will see the volume rendering correctly, so I guess it is possible for quarto to do the same.

My guess is that there is something going on during the formatting of the the ipywidget when adding it to the html file, since I can backengineer the errors in the html file to get it to work.

So basically, the html file I am comparing is one using `jupyter nbconvert --to html FILE.ipynb --execute` and other using `quarto render FILE.ipynb --execute`

**Reply by mcanouil** [reply]
@cscheid in case you have some ideas here.

**Reply by cscheid** [reply]
I don't have any great ideas but I suspect someone is emitting "naked" HTML in the output without a raw block around it, for example:

> For example there is one part of the script that should be ;\"><\/iframe>", but quarto generates ;\"&gt;".

**Reply by cscheid** [reply]
I'm trying to reproduce the failure and I unfortunately can't get this to work. `quarto render` produces the following output in the cell

````
Failed to use notebook backend: 

cannot import name 'vtk' from 'trame.widgets' (/Users/cscheid/virtualenvs/homebrew-python3/lib/python3.10/site-packages/trame/widgets/__init__.py)

Falling back to a static output.
````

I've installed vtk, pyvista, etc. We'll need more help making a reproducible example here.

**Reply by FelipeCybis** [reply]
I will work on this and create a `pyproject.toml`

**Reply by FelipeCybis** [reply]
Hi @cscheid,

You should be good to go with this [`requirements.txt`](https://github.com/FelipeCybis/trame_quarto_example/blob/main/requirements.txt).
I am using quarto `1.4.549`

**Reply by cscheid** [reply]
```
ERROR: Ignored the following versions that require a different python version: 1.21.2 Requires-Python >=3.7,<3.11; 1.21.3 Requires-Python >=3.7,<3.11; 1.21.4 Requires-Python >=3.7,<3.11; 1.21.5 Requires-Python >=3.7,<3.11; 1.21.6 Requires-Python >=3.7,<3.11
ERROR: Could not find a version that satisfies the requirement pywin32==306 (from versions: none)
ERROR: No matching distribution found for pywin32==306
```

I think you're on windows, and so this `requirements.txt` isn't portable?

**Reply by FelipeCybis** [reply]
Seems that I did `pip freeze > requirements.txt` without thinking too hard... let's try this as requirements:
```
ipykernel==6.29.0
ipywidgets==8.1.1
nbconvert==7.14.2
pyvista[all]==0.43.2
```
or, again, [here](https://github.com/FelipeCybis/trame_quarto_example/blob/main/requirements.txt)

**Reply by cscheid** [reply]
Thanks. I see the issue now. I don't think we can do anything about it. The problem, specifically, is that pyvista is emitting a widget state that includes the specific string `</script>`. We are following the recommended approach here: https://ipywidgets.readthedocs.io/en/7.x/embedding.html which is to create a script tag and embed the JSON text.

Unfortunately, this widget state cannot have the `</script>` string, because that causes the outer `<script type="application/vnd.jupyter.widget-state+json">` to close. If ipywidgets allows us to emit this JSON in an escaped way so that it can have the string `</script>` in it, then we can fix the bug on our side (but I can't see a way to do that in the documentation). But unfortunately this looks like something we can't fix on our side.

**Reply by cscheid** [reply]
Hm. There might be a funny hack that works. I'll open an issue.

**Reply by FelipeCybis** [reply]
Great, thanks! Once I have some time I will try to correct this in a fork of pyvista and see if I can propose a PR on their side as well.

**Reply by cscheid** [reply]
There's not much they'll be able to do. The PR would have to be in ipywidgets, to support base64 encoding of the JSON.

**Reply by cscheid** [reply, marked as answer]
Tracking here: https://github.com/quarto-dev/quarto-cli/issues/8433

**Reply by FelipeCybis** [reply]
Thanks for that @cscheid !

Just to see if I understood correctly, if you allow me.

The PR you are opening is adressing the same problem that ipywidgets' utils adresses here https://github.com/jupyter-widgets/ipywidgets/blob/b2531796d414b0970f18050d6819d932417b9953/python/ipywidgets/ipywidgets/embed.py#L230-L241, correct?

I tried your PR branch by the way, and works perfectly :)

**Reply by cscheid** [reply]
Pretty much, yes. If they're escaping their strings explicitly, then I think they should fix the example here to highlight that they expect these strings to be escaped as well: https://ipywidgets.readthedocs.io/en/7.x/embedding.html 

But in any case, this fix is coming soon. We're trying to come up with a test that doesn't take ~10MB of space, but it'll be there shortly. Thanks for confirming!
